src/session.rs

```diff
@@ -0,0 +1,83 @@
+extern crate dirs;
+
+use crate::auth::Auth;
+use crate::request_items::Parameter;
+use crate::utils::ensure_session_dir_exists;
+use serde::{Deserialize, Serialize};
+use std::fs::File;
+use std::io::prelude::*;
+
+const VERSION: &'static str = env!("CARGO_PKG_VERSION");
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Session {
+    pub identifier: String,
+    pub host: String,
+    pub headers: Vec<Parameter>,
+    pub auth: Option<Auth>,
+    ht_version: String,
+}
+
+impl Session {
+    pub fn new(identifier: String, host: &str, headers: Vec<Parameter>, auth: Option<Auth>) -> Session {
+        Session {
+            identifier,
+            host: host.to_string(),
+            headers,
+            auth,
+            ht_version: String::from(VERSION),
+        }
+    }
+
+    pub fn save(&self) -> std::io::Result<String> {
+        let json = serde_json::to_string(&self)?;
+        let identifier = &self.identifier;
+        let mut config_dir = match ensure_session_dir_exists(&self.host) {
+            Err(why) => panic!("couldn't get config directory: {}", why),
+            Ok(dir) => dir,
+        };
+        config_dir.push(identifier);
```

Comment:
IMO it'd be more clear if we did `config_dir.push(&self.identifier)` and in for tail expression `Ok(identifier.clone())` (which'd mean no longer having the `let identifier = &self.identifier` line).

Reply:
True. Besides, the `Result<String>` is no longer needed in this version of the code (I did use the returned identifier in a message at one point), so I replaced it with a `Result<()>`. I hope it makes more sense.